x/txfees/keeper/hooks.go

```diff
@@ -0,0 +1,65 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/osmosis-labs/osmosis/v7/osmoutils"
+	epochstypes "github.com/osmosis-labs/osmosis/v7/x/epochs/types"
+	txfeestypes "github.com/osmosis-labs/osmosis/v7/x/txfees/types"
+)
+
+func (k Keeper) BeforeEpochStart(ctx sdk.Context, epochIdentifier string, epochNumber int64) {}
+
+// at the end of each epoch, swap all non-OSMO fees into OSMO and transfer to fee module account
+func (k Keeper) AfterEpochEnd(ctx sdk.Context, epochIdentifier string, epochNumber int64) {
+	nonNativeFeeAddr := k.accountKeeper.GetModuleAddress(txfeestypes.NonNativeFeeCollectorName)
+	baseDenom, _ := k.GetBaseDenom(ctx)
+	feeTokens := k.GetFeeTokens(ctx)
+
+	for _, feetoken := range feeTokens {
+		if feetoken.Denom == baseDenom {
+			continue
+		}
+		coinBalance := k.bankKeeper.GetBalance(ctx, nonNativeFeeAddr, feetoken.Denom)
+		if coinBalance.Amount.IsZero() {
+			continue
+		}
+
+		// Do the swap of this fee token denom to base denom.
+		_ = osmoutils.ApplyFuncIfNoError(ctx, func(cacheCtx sdk.Context) error {
```

Comment:
should we log the error at least?

Reply:
The error gets logged to ctx in ApplyFuncIfNoError https://github.com/osmosis-labs/osmosis/blob/main/osmoutils/cache_ctx.go#L26